### ISSUE TITLE
[Feature]: publish ARM-based browser binaries

### ISSUE THREAD

**Issue by aslushnikov** [microsoft/playwright#7723]
### Feature request

We currently publish amd64 docker containers. As of July 19, 2021, these don't run on Apple M1 silicon, even with `--platform linux/amd64` flag (see https://github.com/microsoft/playwright/issues/7655).

This is a feature request to publish arm64-based docker images. Upvote to let us know that you're interested in this feature!

**Comment by kumaraditya303** [comment]
This would also cover the use case of running on arm64 raspberry PI 4

**Comment by n8henrie** [comment]
I'm not very experienced with Docker, but after finding that the official images didn't work on M1, I'm surprised to see that I can't even `python -m pip install playwright==1.14.1` on a Debian-based image.

**Comment by n8henrie** [comment]
I am still not having any luck with Docker on an M1 Mac.

```console
$ cat Dockerfile
FROM mcr.microsoft.com/playwright:v1.15.2-focal

RUN python3 -m pip install playwright==1.15.2
ADD test.py /app/

CMD [ "python3", "/app/test.py" ]
$ cat test.py
from playwright.sync_api import sync_playwright


def main():
    with sync_playwright() as p:
        browser = p.firefox.launch()
        page = browser.new_page()
        page.goto("http://playwright.dev")
        print(page.title())
        browser.close()


if __name__ == "__main__":
    main()
$ docker run -it --rm --ipc=host $(docker build -q .)
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Traceback (most recent call last):
  File "/app/test.py", line 14, in <module>
    main()
  File "/app/test.py", line 6, in main
    browser = p.firefox.launch()
  File "/usr/local/lib/python3.8/dist-packages/playwright/sync_api/_generated.py", line 11154, in launch
    self._sync(
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_sync_base.py", line 111, in _sync
    return task.result()
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_browser_type.py", line 88, in launch
    return from_channel(await self._channel.send("launch", params))
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_connection.py", line 39, in send
    return await self.inner_send(method, params, False)
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_connection.py", line 57, in inner_send
    result = next(iter(done)).result()
playwright._impl._api_types.Error: Protocol error (Browser.enable): Browser closed.
==================== Browser output: ====================
<launching> /ms-playwright/firefox-1295/firefox/firefox -no-remote -headless -profile /tmp/playwright_firefoxdev_profile-9vAfUO -juggler-pipe -silent
<launched> pid=565
[pid=565][err] *** You are running in headless mode.
[pid=565][err] [Parent 565, Main Thread] WARNING: failed to seal memfd: Invalid argument: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:461
[pid=565][err] [Parent 565, Main Thread] WARNING: freezeable shared memory was never frozen: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:547
[pid=565][err] qemu: uncaught target signal 11 (Segmentation fault) - core dumped
=========================== logs ===========================
<launching> /ms-playwright/firefox-1295/firefox/firefox -no-remote -headless -profile /tmp/playwright_firefoxdev_profile-9vAfUO -juggler-pipe -silent
<launched> pid=565
[pid=565][err] *** You are running in headless mode.
[pid=565][err] [Parent 565, Main Thread] WARNING: failed to seal memfd: Invalid argument: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:461
[pid=565][err] [Parent 565, Main Thread] WARNING: freezeable shared memory was never frozen: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:547
[pid=565][err] qemu: uncaught target signal 11 (Segmentation fault) - core dumped
============================================================
```

Adding `--platform linux/amd64` doesn't seem to change anything:

```console
$ docker run -it --rm --ipc=host --platform linux/amd64  $(docker build -q .)
Traceback (most recent call last):
  File "/app/test.py", line 14, in <module>
    main()
  File "/app/test.py", line 6, in main
    browser = p.firefox.launch()
  File "/usr/local/lib/python3.8/dist-packages/playwright/sync_api/_generated.py", line 11154, in launch
    self._sync(
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_sync_base.py", line 111, in _sync
    return task.result()
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_browser_type.py", line 88, in launch
    return from_channel(await self._channel.send("launch", params))
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_connection.py", line 39, in send
    return await self.inner_send(method, params, False)
  File "/usr/local/lib/python3.8/dist-packages/playwright/_impl/_connection.py", line 57, in inner_send
    result = next(iter(done)).result()
playwright._impl._api_types.Error: Protocol error (Browser.enable): Browser closed.
==================== Browser output: ====================
<launching> /ms-playwright/firefox-1295/firefox/firefox -no-remote -headless -profile /tmp/playwright_firefoxdev_profile-6W64hM -juggler-pipe -silent
<launched> pid=565
[pid=565][err] *** You are running in headless mode.
[pid=565][err] [Parent 565, Main Thread] WARNING: failed to seal memfd: Invalid argument: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:461
[pid=565][err] [Parent 565, Main Thread] WARNING: freezeable shared memory was never frozen: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:547
[pid=565][err] qemu: uncaught target signal 11 (Segmentation fault) - core dumped
=========================== logs ===========================
<launching> /ms-playwright/firefox-1295/firefox/firefox -no-remote -headless -profile /tmp/playwright_firefoxdev_profile-6W64hM -juggler-pipe -silent
<launched> pid=565
[pid=565][err] *** You are running in headless mode.
[pid=565][err] [Parent 565, Main Thread] WARNING: failed to seal memfd: Invalid argument: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:461
[pid=565][err] [Parent 565, Main Thread] WARNING: freezeable shared memory was never frozen: file /mnt/actions-runner/_work/playwright-internal/playwright-internal/browser_patches/firefox/checkout/ipc/chromium/src/base/shared_memory_posix.cc:547
[pid=565][err] qemu: uncaught target signal 11 (Segmentation fault) - core dumped
============================================================
```

On my Linux machine, it runs without issue:

```console
$ sudo docker run -it --rm --ipc=host $(sudo docker build -q .)
Fast and reliable end-to-end testing for modern web apps | Playwright
```

**Comment by mxschmitt** [comment]
It's not released yet, with version 1.16 it should work.

**Comment by safwanrahman** [comment]
Is there any update on this? Would fixing the CI will trigger the build for previous versions?

**Comment by mxschmitt** [comment]
Current state: We made the Docker images ARM64/Linux compatible. We'll follow-up with browser compatible ARM64/Linux builds in the next release.

So currently its not usable yet, stay tuned!

**Comment by lifesaverluke** [comment]
BTW: In case you need a build specifically for use on AWS Lambda: I created https://www.chromiumforlambda.com

**Comment by aslushnikov** [comment]
Tip-of-tree Playwright now comes with browser builds for Ubuntu 20.04 arm64. It'll be released as 1.17 in upcoming days.

Meanwhile, you can try it out with Playwright Test:

``` 
npm i @playwright/test@next 
```

... or with Playwright library:

```sh
npm i playwright@next
```

**Comment by n8henrie** [comment]
@aslushnikov -- [you suggested](https://github.com/microsoft/playwright/issues/9435#issuecomment-952315565) I follow this issue for an issue I was having with python-playwright -- does this mean the python builds will also be fixed soon (I'm assuming when playwright 1.17 is available in pypi?)

**Comment by edspencer** [comment]
Is this expected to work with the C6G machines types described on https://aws.amazon.com/ec2/instance-types/? They seem to be arm64-based, but when I try to run it using @playwright/test@next I get a bunch of the same errors repeated over and over again:

```[pid=503][err] /home/runner/.cache/ms-playwright/chromium-907428/chrome-linux/chrome: 1: Syntax error: Unterminated quoted string```

This is running inside a docker container using the splendid https://github.com/actions-runner-controller/actions-runner-controller setup. For now I have switched to a Xeon-based EC2 instance and everything works great, would be awesome to use the arm64 machines though as they have a better price/performance curve.

**Comment by kumaraditya303** [comment]
> 📣 Tip-of-tree Playwright now comes with browser builds for Ubuntu 20.04 arm64. **It'll be released as 1.17 in upcoming days**.
> 
> Meanwhile, you can try it out with Playwright Test:
> 
> ``` 
> npm i @playwright/test@next 
> ```
> 
> ... or with Playwright library:
> 
> ```sh
> npm i playwright@next
> ```

@aslushnikov Is arm64 only aimed for nodejs and it is not supported for python? To support python arm64 based drivers also needs to be published.

**Comment by aslushnikov** [comment]
> does this mean the python builds will also be fixed soon (I'm assuming when playwright 1.17 is available in pypi?)

@n8henrie we're looking into this

> Is this expected to work with the C6G machines types described on https://aws.amazon.com/ec2/instance-types/?

@edspencer yes, it should. But you're not running the @next version: in Playwright's @next, chromium should be 939194.

> To support python arm64 based drivers also needs to be published.

@kumaraditya303 yep, thanks! These are published now.

**Comment by dkgrieshammer** [comment]
just as a heads up, 
"@playwright/test": "^1.16.3" & "playwright": "^1.16.3" works on m1, in Dockerized version I had to use @next, resolving to "^1.18.0-alpha-nov-29-2021" to make it work;

**Comment by mxschmitt** [comment]
Yes, it gets released as 1.17.0 tomorrow.
